[AArch64] Change the cost of fma and fmuladd to match fmul.

davemgreen opened the pull request:
As fmul and fmadd are so similar, their performance characteristics tend to be the same on most platforms, at least in terms of reciprocal throughputs. Processors capable of performing a given number of fmul per cycle can usually perform the same number of fma, with the extra add being relatively simple on top. This patch makes the scores of the two operations the same, which brings the throughput cost of a fma/fmuladd to 2, and the latency to 3, which are the defaults for fmul.

Note that we might also want to change the throughput cost of a fmul to 1, as most processors have ample bandwidth for them, but they should still stay in-line with one another.

sjoerdmeijer reviewed (commented):
I just wanted to double check this as there are quite a few of FP multiply accumulate variants. I see they have a latency of 4(2) and throughput of 4 on the Neoverse V2. FMUL has a latency of 3, so FMA/FMULADD doesn't look entirely the same as FMUL, is that right?

davemgreen reviewed (commented):
> I just wanted to double check this as there are quite a few of FP multiply accumulate variants. I see they have a latency of 4(2) and throughput of 4 on the Neoverse V2. FMUL has a latency of 3, so FMA/FMULADD doesn't look entirely the same as FMUL, is that right?

Yeah the latencies will be a little different, but the throughputs are the same. I would not expect a fma/fmuladd to be lower cost though, this is mostly to try and help out #149102 to make sure it doesn't make silly decisions because the default cost of a fma/fmuladd is lower than a fmul.

sjoerdmeijer reviewed (approved):
> this is mostly to try and help out https://github.com/llvm/llvm-project/pull/149102 to make sure it doesn't make silly decisions because the default cost of a fma/fmuladd is lower than a fmul.

Okay, thanks. Yes, I was wondering about the motivation for this, but makes sense. I assume you've checked for potential fall out for this and ran some performance numbers, so LGTM.

davemgreen committed:
[AArch64] Change the cost of fma and fmuladd to match fmul.

I believe that as fmul and fmadd are so similar, their performance
characteristics tend to be the same on most platforms, at least in terms of
reciprical throughputs. Processors capable of performing a given number of fmul
per cycle can usually perform the same number of fma, with the extra add being
relatively simple on top. This makes the scores of the two operations the same,
which brings the throughput cost of a fma/fmuladd to 2, and the latency to 3,
which are the defaults for fmul.

Note that we might also want to change the throughput cost of a fmul to 1, as
most processors have ample bandwodth for them, but they should still stay
in-line with one another.

davemgreen commented:
Hi - I don't tend to run TSVC outside of the llvm-test-suite. It's trying to use a gather, which is going to be pretty slow. It must be just under the cost threshold to vectorize now. The whole thing should be split into two loops, but our loop-distribution is not enabled by default. Or it could load both bits of data and select between them, but the loop distribution would presumably be better in this case.
